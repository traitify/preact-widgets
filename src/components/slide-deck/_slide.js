import { h, Component } from "preact";
import style from "./_slide.scss";

export default class Slide extends Component {
  className (){
    if (this.props.client.oldIE) return style.middleSlide;
    let position;
    if (this.props.slide.orientation == "invisible"){
      position = style.invisible;
    } else if (this.props.slide.orientation == "left"){
      position = style.leftSlide;
    } else if (this.props.slide.orientation == "middle"){
      position = style.middleSlide;
    } else {
      position = style.rightSlide;
    }

    return position;
  }
  style(){
    return {
      backgroundImage: `url(${this.props.slide.image})`,
      backgroundPosition: `${this.props.slide.focus_x}% ${this.props.slide.focus_y}%`
    };
  }
  render() {
    let tag;

    if (!this.props.client.oldIE){
      tag = (
        <div class={`${style.slide} ${this.className()}`} style={this.style()} key={this.id} />
      );
    } else {
      tag = (
        <div class={`${style.slide} ${style.slideIE} ${this.className()}`}>
          <img src={this.props.slide.image} key={this.id} />
        </div>
      );
    }

    return tag;
  }
}
