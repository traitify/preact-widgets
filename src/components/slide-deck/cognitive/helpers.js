import {useEffect, useReducer} from "react";
import {mutable} from "../helpers";

export * from "../helpers";

function loadImage({dispatch, imageType, url}) {
  dispatch({type: "imageLoading"});

  const img = document.createElement("img");
  img.onload = () => dispatch({imageType, type: "imageLoaded"});
  img.onerror = () => {
    // TODO: Delay?
    // TODO: Retries?
    dispatch({error: "Image could not load", type: "error"});
  };
  img.src = url;
}

function loadQuestions(questions, dispatch) {
  const question = questions.find(({loaded}) => !loaded);
  if(!question || !question.image) { return; }
  if(!question.image.loaded) {
    loadImage({dispatch, imageType: "question", url: question.image.url});
    return;
  }

  const response = question.responses.find(({image: {loaded}}) => !loaded);
  if(response) {
    loadImage({dispatch, imageType: "response", url: response.image.url});
  }
}

function reducer(state, action) {
  switch(action.type) {
    case "error":
      return {...state, error: action.error};
    case "imageLoaded": {
      const questions = mutable(state.questions);
      const questionIndex = questions.findIndex(({loaded}) => !loaded);
      const question = questions[questionIndex];

      if(action.imageType === "question") {
        question.image.loaded = true;
      } else {
        const responseIndex = question.responses.findIndex(({image: {loaded}}) => !loaded);

        question.responses[responseIndex].image.loaded = true;
      }

      const responseIndex = question.responses.findIndex(({image: {loaded}}) => !loaded);
      const responseImagesLoaded = responseIndex === -1;

      question.loaded = question.image.loaded && responseImagesLoaded;
      questions[questionIndex] = question;

      const loading = questions.findIndex(({loaded}) => !loaded) !== -1;

      return {...state, error: null, imageLoading: false, loading, questions};
    }
    case "imageLoading":
      return {...state, imageLoading: true, loading: true};
    case "reset":
      return {questions: action.questions};
    case "response": {
      const questions = mutable(state.questions);
      questions[action.questionIndex].answer = action.answer;

      return {...state, error: null, questions};
    }
    default:
      return {...state};
  }
}

export function useQuestionsLoader(initialQuestions) {
  const [
    {error, imageLoading, loading, questions},
    dispatch
  ] = useReducer(reducer, {questions: initialQuestions});

  useEffect(() => dispatch({questions: initialQuestions, type: "reset"}), [initialQuestions]);
  useEffect(() => {
    if(!imageLoading) { loadQuestions(questions, dispatch); }
  }, [imageLoading, questions]);

  return {error, dispatch, loading, questions};
}
