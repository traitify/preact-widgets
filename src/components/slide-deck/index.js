import PropTypes from "prop-types";
import {Component} from "react";
import TraitifyPropTypes from "lib/helpers/prop-types";
import withTraitify from "lib/with-traitify";
import {
  completedSlides,
  dataChanged,
  getStateFromProps,
  isFinished,
  isFullscreen,
  isReady,
  loadingIndex,
  mutable,
  slideIndex,
  toggleFullscreen
} from "./helpers";
import Loading from "./loading";
import Slide from "./slide";
import style from "./style.scss";

class SlideDeck extends Component {
  static defaultProps = {assessment: null}
  static propTypes = {
    assessment: PropTypes.shape({
      id: PropTypes.string.isRequired,
      instructions: PropTypes.shape({instructional_text: PropTypes.string.isRequired}),
      locale_key: PropTypes.string.isRequired,
      slides: PropTypes.arrayOf(PropTypes.object).isRequired
    }),
    assessmentID: PropTypes.string.isRequired,
    cache: PropTypes.shape({
      get: PropTypes.func.isRequired,
      set: PropTypes.func.isRequired
    }).isRequired,
    getAssessment: PropTypes.func.isRequired,
    getOption: PropTypes.func.isRequired,
    isReady: PropTypes.func.isRequired,
    traitify: TraitifyPropTypes.traitify.isRequired,
    translate: PropTypes.func.isRequired,
    ui: TraitifyPropTypes.ui.isRequired
  }
  constructor(props) {
    super(props);

    this.state = {
      isFullscreen: false,
      ...getStateFromProps(props)
    };
  }
  componentDidMount() {
    this.props.ui.trigger("SlideDeck.initialized", this);

    if(isFinished(this.state.slides)) {
      this.finish();
    } else if(this.state.slides.length > 0) {
      this.resizeImages();
    }

    window.addEventListener("resize", this.resizeImages);
    window.addEventListener("fullscreenchange", this.fullscreenToggled);
    window.addEventListener("webkitfullscreenchange", this.fullscreenToggled);
    window.addEventListener("mozfullscreenchange", this.fullscreenToggled);
    window.addEventListener("MSFullscreenChange", this.fullscreenToggled);
  }
  componentDidUpdate(prevProps) {
    this.props.ui.trigger("SlideDeck.updated", this);

    if(dataChanged(this.props.assessment, prevProps.assessment)) {
      this.setState(getStateFromProps(this.props), () => {
        if(isFinished(this.state.slides)) {
          this.finish();
        } else if(this.state.slides.length > 0) {
          this.resizeImages();
        }
      });
    }
  }
  componentWillUnmount() {
    window.removeEventListener("resize", this.resizeImages);
    window.removeEventListener("fullscreenchange", this.fullscreenToggled);
    window.removeEventListener("webkitfullscreenchange", this.fullscreenToggled);
    window.removeEventListener("mozfullscreenchange", this.fullscreenToggled);
    window.removeEventListener("MSFullscreenChange", this.fullscreenToggled);
  }
  fetchImages = () => {
    if(this.state.imageLoading) { return; }

    const currentIndex = loadingIndex(this.state.slides);
    if(currentIndex === -1) { return; }
    if(this.state.imageLoadingAttempts >= 3) { return; }

    this.setState({imageLoading: true}, () => {
      const slide = this.state.slides[currentIndex];
      const img = document.createElement("img");
      img.src = slide.image;
      img.onload = () => {
        this.setState((state) => {
          const slides = mutable(state.slides);
          slides[currentIndex].loaded = true;

          return {
            imageLoading: false,
            imageLoadingAttempts: 0,
            slides
          };
        }, this.fetchImages);
      };
      img.onerror = () => {
        this.setState((state) => (
          {imageLoading: false, imageLoadingAttempts: state.imageLoadingAttempts + 1}
        ), () => {
          setTimeout(this.fetchImages, 2000);
        });
      };
    });
  }
  fullscreenToggled = () => {
    this.setState({isFullscreen: isFullscreen()}, () => {
      this.props.ui.trigger("SlideDeck.fullscreen", this, this.state.isFullscreen);
      this.resizeImages();
    });
  }
  resizeImages = () => {
    this.setState((state, props) => {
      if(!this.container) { return; }

      const width = this.container.clientWidth;
      const height = this.container.clientHeight;
      const imageHost = props.getOption("imageHost");
      const slides = state.slides.map((slide) => ({
        ...slide,
        loaded: false,
        image: width > 0 && height > 0
          ? `${imageHost}/slides/${slide.image_desktop_retina.split("/").pop()}?w=${width}&h=${height}&fp-x=${slide.focus_x / 100}&fp-y=${slide.focus_y / 100}&auto=format`
          : slide.image_desktop
      }));

      return {imageLoadingAttempts: 0, slides};
    }, this.fetchImages);
  }
  finish() {
    if(this.state.finished) { return; }
    this.setState({finished: true});

    if(this.props.isReady("results")) { return; }

    // TODO: If error, display it
    this.props.traitify.put(`/assessments/${this.props.assessmentID}/slides`, completedSlides(this.state.slides)).then((response) => {
      this.props.ui.trigger("SlideDeck.finished", this, response);
      this.props.getAssessment({force: true});
    });
  }
  // Event Methods
  back = () => {
    this.setState((state) => {
      const slides = mutable(state.slides);

      slides[slideIndex(slides) - 1].response = null;

      return {slides, startTime: Date.now()};
    }, this.fetchImages);
  }
  hideInstructions = () => {
    this.setState({showInstructions: false});
  }
  retry = () => {
    this.setState({imageLoadingAttempts: 0}, this.fetchImages);
  }
  toggleFullscreen = () => {
    toggleFullscreen({current: this.state.isFullscreen, element: this.container});
  }
  updateSlide = (index, value) => {
    this.setState((state) => {
      const key = value ? "me" : "notMe";
      this.props.ui.trigger(`SlideDeck.${key}`, this);
      this.props.ui.trigger("SlideDeck.updateSlide", this);

      const slides = mutable(state.slides);
      slides[index].response = value;
      slides[index].time_taken = Date.now() - state.startTime;

      return {slides, startTime: Date.now()};
    }, () => {
      const {assessmentID, cache} = this.props;

      cache.set(`slides.${assessmentID}`, completedSlides(this.state.slides));

      if(isFinished(this.state.slides)) { this.finish(); }
    });
  }
  render() {
    if(this.props.isReady("results")) { return null; }

    const finished = isFinished(this.state.slides);
    const ready = isReady(this.state.slides);

    return (
      <div className={style.widgetContainer} ref={(container) => { this.container = container; }}>
        {(finished || !ready) ? (
          <Loading
            imageLoading={this.state.imageLoadingAttempts < 3}
            retry={this.retry}
            translate={this.props.translate}
          />
        ) : (
          <Slide
            back={this.back}
            getOption={this.props.getOption}
            instructions={this.state.instructions}
            isComplete={finished}
            isFullscreen={this.state.isFullscreen}
            showInstructions={this.state.showInstructions}
            slideIndex={slideIndex(this.state.slides)}
            slides={this.state.slides}
            start={this.hideInstructions}
            toggleFullscreen={this.toggleFullscreen}
            translate={this.props.translate}
            updateSlide={this.updateSlide}
          />
        )}
      </div>
    );
  }
}

export {SlideDeck as Component};
export default withTraitify(SlideDeck);
