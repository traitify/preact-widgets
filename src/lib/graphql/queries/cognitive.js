import toQuery from "../to-query";

const defaultFields = [
  "allottedTime",
  "completed",
  "id",
  "learningDisability",
  "localeKey",
  "name",
  "percentileScore",
  "rawScore",
  "specialAllottedTime",
  "surveyId",
  "surveyKey",
  {
    questions: [
      "id",
      {
        answer: ["answerId", "skipped", "timeTaken"],
        questionImage: ["id", "url"],
        responses: ["id", {image: ["id", "url"]}]
      }
    ]
  }
];

export function create({fields, params}) {
  return {
    query: `
      mutation($localeKey: String, $surveyId: String!) {
        createCognitiveTest(localeKey: $localeKey, surveyId: $surveyId) {
          ${toQuery(fields || defaultFields)}
        }
      }
    `,
    variables: params
  };
}

export function get({fields, params}) {
  return {
    query: `
      query cognitiveTest($localeKey: String, $testId: String!) {
        cognitiveTest(localeKey: $localeKey, testId: $testId) {
          ${toQuery(fields || defaultFields)}
        }
      }
    `,
    variables: params
  };
}

export function surveys(options) {
  const fields = options && options.fields;

  return {
    query: `
      query {
        cognitiveSurveys {
          ${toQuery(fields || {edges: {node: ["id", "key", "name"]}})}
        }
      }
    `
  };
}

export function update({fields, params}) {
  return {
    query: `
      mutation(
        $answers: [QuestionAnswer]!,
        $learningDisability: Boolean,
        $overallTimeTaken: Int!,
        $testId: String!
      ) {
        completeCognitiveTest(
          answers: $answers,
          learningDisability: $learningDisability,
          overallTimeTaken: $overallTimeTaken,
          testId: $testId
        ) {
          ${toQuery(fields || ["message", "success"])}
        }
      }
    `,
    variables: params
  };
}
