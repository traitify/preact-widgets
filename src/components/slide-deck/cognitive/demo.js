import PropTypes from "prop-types";
import {useEffect, useState} from "react";
import Loading from "components/loading";
import Slide from "./slide";
import {useSlidesLoader} from "./helpers";

const defaultSlides = [
  {
    id: "s-1",
    questionImage: {url: "https://via.placeholder.com/300?text=Test+Question+1"},
    responses: [
      {id: "r-1", image: {url: "https://via.placeholder.com/100/008000?text=Response+A"}},
      {id: "r-2", image: {url: "https://via.placeholder.com/100/0000FF?text=Response+B"}},
      {id: "r-3", image: {url: "https://via.placeholder.com/100/FFFF00?text=Response+C"}},
      {id: "r-4", image: {url: "https://via.placeholder.com/100/FF0000?text=Response+D"}}
    ]
  },
  {
    id: "s-2",
    questionImage: {url: "https://via.placeholder.com/300?text=Test+Question+2"},
    responses: [
      {id: "r-5", image: {url: "https://via.placeholder.com/100/0000FF?text=Response+A"}},
      {id: "r-6", image: {url: "https://via.placeholder.com/100/FFFF00?text=Response+B"}},
      {id: "r-7", image: {url: "https://via.placeholder.com/100/FF0000?text=Response+C"}},
      {id: "r-8", image: {url: "https://via.placeholder.com/100/008000?text=Response+D"}}
    ]
  },
  {
    id: "s-3",
    questionImage: {url: "https://via.placeholder.com/300?text=Test+Question+3"},
    responses: [
      {id: "r-9", image: {url: "https://via.placeholder.com/100/FFFF00?text=Response+A"}},
      {id: "r-10", image: {url: "https://via.placeholder.com/100/FF0000?text=Response+B"}},
      {id: "r-11", image: {url: "https://via.placeholder.com/100/008000?text=Response+C"}},
      {id: "r-12", image: {url: "https://via.placeholder.com/100/0000FF?text=Response+D"}}
    ]
  },
  {
    id: "s-4",
    questionImage: {url: "https://via.placeholder.com/300?text=Test+Question+4"},
    responses: [
      {id: "r-13", image: {url: "https://via.placeholder.com/100/FF0000?text=Response+A"}},
      {id: "r-14", image: {url: "https://via.placeholder.com/100/008000?text=Response+B"}},
      {id: "r-15", image: {url: "https://via.placeholder.com/100/0000FF?text=Response+C"}},
      {id: "r-16", image: {url: "https://via.placeholder.com/100/FFFF00?text=Response+D"}}
    ]
  }
];

function Demo(props) {
  const [initialSlides, setInitialSlides] = useState([]);
  const [slideIndex, setSlideIndex] = useState(0);
  const {dispatch, error, slides} = useSlidesLoader(initialSlides);
  const onSelect = (answer) => {
    dispatch({answer, slideIndex, type: "response"});
    setSlideIndex(slideIndex + 1);
  };

  useEffect(() => {
    // TODO: Get from API
    setInitialSlides(defaultSlides);
  }, []);

  useEffect(() => {
    if(slides.length === 0) { return; }
    if(slides.length !== slideIndex) { return; }

    props.onFinish();
  }, [slideIndex]);

  const slide = slides[slideIndex];

  if(!slide) { return <Loading />; }

  // TODO: Display error?
  // TODO: Retry?
  if(error) { console.log(error); }

  return (
    <div>
      <Slide onSelect={onSelect} slide={slides[slideIndex]} />
    </div>
  );
}

Demo.propTypes = {
  onFinish: PropTypes.func.isRequired
};

export default Demo;
