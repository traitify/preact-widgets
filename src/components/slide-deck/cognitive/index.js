/* eslint-disable no-alert */
/* eslint-disable no-console */
// TODO: Remove no-console line
import {faClock} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
import {useEffect, useState} from "react";
import Loading from "components/loading";
import Icon from "lib/helpers/icon";
import withTraitify from "lib/with-traitify";
import Instructions from "./instructions";
import Slide from "./slide";
import {useQuestionsLoader} from "./helpers";
import style from "./style.scss";

function Cognitive(props) {
  const disableTimeLimit = props.getOption("disableTimeLimit");
  const [initialQuestions, setInitialQuestions] = useState([]);
  const {dispatch, questions} = useQuestionsLoader(initialQuestions);
  const [disability, setDisability] = useState(false);
  const [onlySkipped, setOnlySkipped] = useState(false);
  const [questionIndex, setQuestionIndex] = useState(null);
  const [skipped, setSkipped] = useState(null);
  const [startTime, setStartTime] = useState(null);
  const [submitting, setSubmitting] = useState(false);
  const [timeLeft, setTimeLeft] = useState(null);
  const onSelect = (answer) => {
    dispatch({answer, questionIndex, type: "response"});
    if(!onlySkipped) { return setQuestionIndex(questionIndex + 1); }

    const newQuestionIndex = questions
      .findIndex((question, index) => index > questionIndex && !question.answer.answerId);
    setQuestionIndex(newQuestionIndex === -1 ? questions.length + 1 : newQuestionIndex);
  };
  const onStart = (options) => {
    if(options.disability) { setDisability(true); }

    setQuestionIndex(0);
    setStartTime(Date.now());
  };
  const onSubmit = () => {
    if(submitting) { return; }

    setSubmitting(true);
    setTimeLeft(0);
    console.log("Submitting", questions);
    // TODO: Ajax
  };

  useEffect(() => {
    if(!props.assessment) { return; }
    if(!props.assessment.questions) { return; }

    const cachedData = props.cache.get(`cognitive.${props.assessment.id}`);
    if(cachedData) {
      setDisability(cachedData.disability);
      setOnlySkipped(cachedData.onlySkipped);
      setSkipped(cachedData.skipped);

      setQuestionIndex(cachedData.questionIndex);
      setInitialQuestions(cachedData.questions);
      setStartTime(cachedData.startTime);
    } else {
      setInitialQuestions(props.assessment.questions);
    }
  }, [
    props.assessment && props.assessment.questions,
    props.assessment && props.assessment.localeKey
  ]);

  useEffect(() => {
    if(!props.assessment) { return; }
    if(questions.length === 0) { return; }

    props.cache.set(`cognitive.${props.assessment.id}`, {
      disability,
      onlySkipped,
      questions,
      questionIndex,
      skipped,
      startTime
    });
  }, [disability, onlySkipped, questions, questionIndex, skipped, startTime]);

  useEffect(() => {
    if(disableTimeLimit) { return; }
    if(!startTime) { return; }
    if(submitting) { return; }

    const calculateTimeLeft = () => {
      const allowedTime = 1000 * (disability ? 385 : 300);
      const timePassed = Date.now() - startTime;
      const _timeLeft = (allowedTime - timePassed) / 1000;

      return _timeLeft < 0 ? 0 : _timeLeft;
    };

    if(!timeLeft) { return setTimeLeft(calculateTimeLeft()); }

    setTimeout(() => {
      const newTimeLeft = calculateTimeLeft();

      setTimeLeft(newTimeLeft);
    }, 1000);
  }, [startTime, timeLeft]);

  useEffect(() => {
    if(disableTimeLimit) { return; }
    if(timeLeft === 0) { onSubmit(); }
  }, [timeLeft]);

  useEffect(() => {
    if(questions.length === 0) { return; }
    if(questions.length > questionIndex) { return; }
    if(onlySkipped) { return onSubmit(); }

    setOnlySkipped(true);

    const skippedIndexes = questions
      .map(({answer}, index) => ({answer, index}))
      .filter(({answer}) => !answer.answerId)
      .map(({index}) => index);

    if(skippedIndexes.length === 0) { return onSubmit(); }
    if(window.confirm("You have some additional time, would you like to attempt the questions you skipped?")) {
      setSkipped(skippedIndexes);
      setQuestionIndex(skippedIndexes[0]);
    } else {
      return onSubmit();
    }
  }, [questions, questionIndex]);

  const question = questions[questionIndex];

  if(questionIndex === null) { return <Instructions onStart={onStart} />; }
  if(!question || submitting) { return <Loading />; }

  const minutes = Math.floor((timeLeft / 60) % 60);
  const seconds = Math.floor(timeLeft % 60);
  const index = skipped ? skipped.indexOf(questionIndex) : questionIndex;
  const total = skipped ? skipped.length : questions.length;
  const progress = 100.0 * (index + 1) / total;

  return (
    <div className={style.container}>
      <div className={style.statusContainer}>
        {!disableTimeLimit && (
          <div className={style.timer}>
            <Icon icon={faClock} />
            &nbsp; {minutes}:{seconds < 10 ? `0${seconds}` : seconds}
          </div>
        )}
        <div className={style.status}>
          {skipped && <span>Skipped Questions </span>}
          <span>{index + 1} / {total}</span>
        </div>
        <div className={style.progressBar}>
          <div className={style.progress} style={{width: `${progress}%`}} />
        </div>
      </div>
      <Slide onSelect={onSelect} question={question} />
    </div>
  );
}

Cognitive.defaultProps = {assessment: null};
Cognitive.propTypes = {
  assessment: PropTypes.shape({
    id: PropTypes.string.isRequired,
    localeKey: PropTypes.string.isRequired,
    questions: PropTypes.arrayOf(PropTypes.object).isRequired
  }),
  cache: PropTypes.shape({
    get: PropTypes.func.isRequired,
    set: PropTypes.func.isRequired
  }).isRequired,
  getOption: PropTypes.func.isRequired
};

export {Cognitive as Component};
export default withTraitify(Cognitive);
