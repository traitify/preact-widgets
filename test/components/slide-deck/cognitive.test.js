import {Component} from "components/slide-deck/cognitive";
import Instructions from "components/slide-deck/cognitive/instructions";
import Slide from "components/slide-deck/cognitive/slide";
import ComponentHandler from "support/component-handler";
import assessment from "support/json/assessment/cognitive.json";

jest.mock("components/slide-deck/cognitive/instructions", () => (() => <div className="mock">Instructions</div>));
jest.mock("components/slide-deck/cognitive/slide", () => (() => <div className="mock">Slide</div>));
jest.mock("components/slide-deck/cognitive/timer", () => (() => <div className="mock">4:59</div>));
jest.mock("lib/with-traitify", () => ((value) => value));

describe("Cognitive", () => {
  const cacheKey = `cognitive.slide-deck.${assessment.id}`;
  let originalConfirm;
  let props;
  const lastUpdate = () => props.ui.trigger.mock.calls[props.ui.trigger.mock.calls.length - 1][1];

  beforeEach(() => {
    originalConfirm = window.confirm;
    window.confirm = jest.fn().mockName("confirm");

    props = {
      assessment,
      cache: {
        get: jest.fn().mockName("get"),
        set: jest.fn().mockName("set")
      },
      getCognitiveAssessment: jest.fn().mockName("getCognitiveAssessment"),
      getOption: jest.fn().mockName("getOption"),
      isReady: jest.fn().mockName("isReady").mockImplementation((value) => value === "questions"),
      traitify: {
        get: jest.fn().mockName("get"),
        post: jest.fn().mockName("post").mockResolvedValue(Promise.resolve()),
        put: jest.fn().mockName("put")
      },
      translate: jest.fn().mockName("translate").mockImplementation((value) => value),
      ui: {
        current: {},
        off: jest.fn().mockName("off"),
        on: jest.fn().mockName("on"),
        trigger: jest.fn().mockName("trigger")
      }
    };
  });

  afterEach(() => {
    window.confirm = originalConfirm;
  });

  describe("callbacks", () => {
    it("triggers initialization", () => {
      new ComponentHandler(<Component {...props} />);

      expect(props.ui.trigger).toHaveBeenNthCalledWith(
        1,
        "SlideDeck.initialized",
        expect.objectContaining({
          props: expect.any(Object),
          state: expect.any(Object)
        })
      );
    });

    it("triggers update", () => {
      new ComponentHandler(<Component {...props} />);

      expect(props.ui.trigger).toHaveBeenNthCalledWith(
        2,
        "SlideDeck.updated",
        expect.objectContaining({
          props: expect.any(Object),
          state: expect.any(Object)
        })
      );
    });
  });

  describe("setup", () => {
    it("requires an assessment", () => {
      props.assessment = null;

      new ComponentHandler(<Component {...props} />);

      expect(props.cache.get).not.toHaveBeenCalled();
    });

    it("sets everything from cache", () => {
      const state = {
        disability: true,
        onlySkipped: true,
        questions: [assessment.questions[0]],
        questionIndex: 1,
        skipped: true,
        startTime: Date.now()
      };
      props.cache.get.mockReturnValue(state);

      new ComponentHandler(<Component {...props} />);

      const lastState = lastUpdate().state;

      expect(props.cache.get).toHaveBeenCalled();
      expect(lastState.disability).toBe(state.disability);
      expect(lastState.initialQuestions).toBe(state.questions);
      expect(lastState.onlySkipped).toBe(state.onlySkipped);
      expect(lastState.questionIndex).toBe(state.questionIndex);
      expect(lastState.skipped).toBe(state.skipped);
      expect(lastState.startTime).toBe(state.startTime);
    });

    it("sets questions from props", () => {
      new ComponentHandler(<Component {...props} />);

      expect(props.cache.get).toHaveBeenCalled();
      expect(lastUpdate().state.initialQuestions).toBe(assessment.questions);
    });
  });

  it("renders instructions", () => {
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders loading if questions finished", () => {
    props.assessment = {...assessment, questions: props.assessment.questions.slice(0, 1)};
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: false}));
    const slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[0].id,
      timeTaken: 600
    }));

    expect(component.tree).toMatchSnapshot();
  });

  it("renders loading after skipped slides finished", () => {
    window.confirm.mockReturnValue(true);
    props.assessment = {...assessment, questions: props.assessment.questions.slice(0, 3)};
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: false}));
    let slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[0].id,
      timeTaken: 600
    }));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({skipped: true, timeTaken: 600}));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({skipped: true, timeTaken: 600}));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[1].id,
      timeTaken: 600
    }));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[2].id,
      timeTaken: 600
    }));

    expect(component.tree).toMatchSnapshot();
    expect(window.confirm).toHaveBeenCalledTimes(1);
    expect(props.cache.set).toHaveBeenLastCalledWith(cacheKey, {
      disability: false,
      onlySkipped: true,
      questions: expect.any(Array),
      questionIndex: 4,
      skipped: [1, 2],
      startTime: expect.any(Number)
    });
  });

  it("renders loading after skipped slides skipped", () => {
    window.confirm.mockReturnValue(false);
    props.assessment = {...assessment, questions: props.assessment.questions.slice(0, 3)};
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: false}));
    let slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[0].id,
      timeTaken: 600
    }));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({skipped: true, timeTaken: 600}));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({skipped: true, timeTaken: 600}));

    expect(component.tree).toMatchSnapshot();
    expect(window.confirm).toHaveBeenCalledTimes(1);
    expect(props.cache.set).toHaveBeenLastCalledWith(cacheKey, {
      disability: false,
      onlySkipped: true,
      questions: expect.any(Array),
      questionIndex: 3,
      skipped: null,
      startTime: expect.any(Number)
    });
  });

  it("renders nothing if results ready", () => {
    props.isReady.mockImplementation(() => true);
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
    expect(() => component.instance.findByType(Slide)).toThrow();
  });

  it("renders skipped slides", () => {
    window.confirm.mockReturnValue(true);
    props.assessment = {...assessment, questions: props.assessment.questions.slice(0, 3)};
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: false}));
    let slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[0].id,
      timeTaken: 600
    }));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({skipped: true, timeTaken: 600}));
    slide = component.instance.findByType(Slide);
    component.act(() => slide.props.onSelect({
      answerId: slide.props.question.responses[1].id,
      timeTaken: 600
    }));

    expect(component.tree).toMatchSnapshot();
    expect(window.confirm).toHaveBeenCalledTimes(1);
    expect(props.cache.set).toHaveBeenLastCalledWith(cacheKey, {
      disability: false,
      onlySkipped: true,
      questions: expect.any(Array),
      questionIndex: 1,
      skipped: [1],
      startTime: expect.any(Number)
    });
  });

  it("renders slide", () => {
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: false}));

    expect(component.tree).toMatchSnapshot();
    expect(() => component.instance.findByType(Instructions)).toThrow();
    expect(props.cache.set).toHaveBeenLastCalledWith(cacheKey, {
      disability: false,
      onlySkipped: false,
      questions: expect.any(Array),
      questionIndex: 0,
      skipped: null,
      startTime: expect.any(Number)
    });
  });

  it("renders slide with disability", () => {
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: true}));

    expect(component.tree).toMatchSnapshot();
    expect(() => component.instance.findByType(Instructions)).toThrow();
    expect(props.cache.set).toHaveBeenLastCalledWith(cacheKey, {
      disability: true,
      onlySkipped: false,
      questions: expect.any(Array),
      questionIndex: 0,
      skipped: null,
      startTime: expect.any(Number)
    });
  });

  it("renders slide with time limit disabled", () => {
    props.getOption.mockImplementation((value, nestedValue) => value === "slideDeck" && nestedValue === "disableTimeLimit");
    const component = new ComponentHandler(<Component {...props} />);
    const instructions = component.instance.findByType(Instructions);
    component.act(() => instructions.props.onStart({disability: true}));

    expect(component.tree).toMatchSnapshot();
    expect(() => component.instance.findByType(Instructions)).toThrow();
    expect(props.cache.set).toHaveBeenLastCalledWith(cacheKey, {
      disability: true,
      onlySkipped: false,
      questions: expect.any(Array),
      questionIndex: 0,
      skipped: null,
      startTime: expect.any(Number)
    });
  });
});
