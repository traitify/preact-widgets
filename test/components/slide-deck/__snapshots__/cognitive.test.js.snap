// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cognitive renders instructions 1`] = `
<div
  className="mock"
>
  Instructions
</div>
`;

exports[`Cognitive renders loading after skipped slides finished 1`] = `
<div>
  <div>
    <i />
    <i />
  </div>
</div>
`;

exports[`Cognitive renders loading after skipped slides skipped 1`] = `
<div>
  <div>
    <i />
    <i />
  </div>
</div>
`;

exports[`Cognitive renders loading if questions finished 1`] = `
<div>
  <div>
    <i />
    <i />
  </div>
</div>
`;

exports[`Cognitive renders nothing if results ready 1`] = `null`;

exports[`Cognitive renders skipped slides 1`] = `
<div>
  <div>
    <div
      className="mock"
    >
      4:59
    </div>
    <div>
      <span>
        cognitive_skipped_questions
         
      </span>
      <span>
        1
         / 
        1
      </span>
    </div>
    <div>
      <div
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Cognitive renders slide 1`] = `
<div>
  <div>
    <div
      className="mock"
    >
      4:59
    </div>
    <div>
      <span>
        1
         / 
        30
      </span>
    </div>
    <div>
      <div
        style={
          Object {
            "width": "3.3333333333333335%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Cognitive renders slide with disability 1`] = `
<div>
  <div>
    <div
      className="mock"
    >
      4:59
    </div>
    <div>
      <span>
        1
         / 
        30
      </span>
    </div>
    <div>
      <div
        style={
          Object {
            "width": "3.3333333333333335%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;

exports[`Cognitive renders slide with time limit disabled 1`] = `
<div>
  <div>
    <div>
      <span>
        1
         / 
        30
      </span>
    </div>
    <div>
      <div
        style={
          Object {
            "width": "3.3333333333333335%",
          }
        }
      />
    </div>
  </div>
  <div
    className="mock"
  >
    Slide
  </div>
</div>
`;
